### PR TITLE
Allow users to preview image in attachment list in the task detail modal

### PR DESCRIPTION
**Pull request by Travisliang001**
# Preview Image Demo
image
image

[(task:23035)](https://staging.teamyapp.com/teams/107/tasks/23035)
# Scope

- Allow users to preview attachments(image only)
- The icons in the list are hoverable and indicate they are clickable for previewing content.
- Set a fixed size for the preview window; if the image exceeds this size, enable scrolling within the preview window.

### CODE REVIEW

File: src/components/internal/Modals/TaskDetailModal.component.tsx [hunk 1]
```diff
@@ -365,6 +371,10 @@ export class TaskDetailModalComponent extends Component<Props, State> {
                         taskId={this.state.task?.id}
                     />
                 )}
+                <PreviewAttachemntComponent
+                    ref={this.previewAttachmentModalRef}
+                    attachmentUrl={this.state?.previewAttachmentUrl}
```
Comment (duanxianbo) [inline src/components/internal/Modals/TaskDetailModal.component.tsx]:
```suggestion
                    attachmentUrl={this.state.previewAttachmentUrl}
```

File: src/components/internal/PreviewAttachment.component.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                    className={styles.CloseButton}
+                    onClick={modalRef.current?.close}
+                >
+                    <MaterialIconUI>cancel</MaterialIconUI>
+                </div>
+            </div>
+            <div className={styles.ModalContent}>
+                <img src={props.attachmentUrl} alt="Preview" />
+            </div>
+        </ModalUI>
```
Comment (duanxianbo) [inline src/components/internal/PreviewAttachment.component.tsx]: can we format this?

File: src/components/internal/Modals/TaskDetailModal.component.tsx (hunk 1 above)
Comment (duanxianbo) [reply src/components/internal/Modals/TaskDetailModal.component.tsx]: I think this component may be better to live within AttachmentList

**Comment by duanxianbo** [comment]
will we create button to allow user to go next and prev previews?

File: src/components/internal/PreviewAttachment.component.tsx (hunk 2 above)
Comment (duanxianbo) [reply src/components/internal/PreviewAttachment.component.tsx]: yarn format

File: src/components/internal/Modals/TaskDetailModal.component.tsx (hunk 1 above)
Comment (naturetime10) [reply src/components/internal/Modals/TaskDetailModal.component.tsx]: Can we pass in `attachmentUrl` and `attachmentType` through `open` method of `previewAttachmentModalRef`?

File: src/components/internal/Modals/TaskDetailModal.component.tsx [hunk 3]
```diff
@@ -615,4 +625,11 @@ export class TaskDetailModalComponent extends Component<Props, State> {
     private onDeleteTaskLinkClick = (linkId: number) => () => {
         this.stateSyncer.deleteTaskLink(linkId);
     };
+
+    private onPreviewAttachmentClick = (url: string) => {
```
Comment (naturetime10) [inline src/components/internal/Modals/TaskDetailModal.component.tsx]:
```suggestion
    private onPreviewAttachmentClick = (attachmentType: AttachmentType, url: string) => {
```

File: src/components/internal/Modals/TaskDetailModal.component.tsx [hunk 4]
```diff
@@ -615,4 +625,11 @@ export class TaskDetailModalComponent extends Component<Props, State> {
...
+        this.setState({
+            previewAttachmentUrl: url,
+        });
```
Comment (naturetime10) [inline src/components/internal/Modals/TaskDetailModal.component.tsx]:
```suggestion
```

File: src/components/internal/Modals/TaskDetailModal.component.tsx [hunk 5]
```diff
@@ -615,4 +625,11 @@ export class TaskDetailModalComponent extends Component<Props, State> {
...
+        this.previewAttachmentModalRef.current?.open();
```
Comment (naturetime10) [inline src/components/internal/Modals/TaskDetailModal.component.tsx]:
```suggestion
        this.previewAttachmentModalRef.current?.open(attachmentType, url);
```

File: src/components/internal/PreviewAttachment.component.module.scss
Comment (naturetime10) [inline src/components/internal/PreviewAttachment.component.module.scss:1]: Can we rename `PreviewAttachmentModal.component.module.scss`.

File: src/components/internal/PreviewAttachment.component.tsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import {
+    ForwardRefExoticComponent,
+    ForwardedRef,
+    forwardRef,
+    useImperativeHandle,
+    useRef,
+} from 'react';
+import { MaterialIconUI } from '@lib/ui/MaterialIcon';
+import styles from './PreviewAttachment.component.module.scss';
+
```
Comment (naturetime10) [inline src/components/internal/PreviewAttachment.component.tsx]:
```suggestion
```

File: src/components/internal/PreviewAttachment.component.tsx
Comment (naturetime10) [inline src/components/internal/PreviewAttachment.component.tsx:1]: Can we rename this file to `src/components/internal/PreviewAttachmentModal.component.tsx` and move into `Modals`?

File: src/components/internal/PreviewAttachment.component.tsx [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        const open = () => {
+            modalRef.current?.open();
+        };
+        useImperativeHandle(ref, () => ({
+            open,
+        }));
+
+        return <ModalUI ref={modalRef} level={2}>
+            <div className={styles.Header}>
+                Image Preview
```
Comment (naturetime10) [inline src/components/internal/PreviewAttachment.component.tsx]:
```suggestion
               Preview Attachment
```

File: src/components/internal/PreviewAttachment.component.tsx [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+                    role={'button'}
+                    aria-label={'Close'}
+                    className={styles.CloseButton}
+                    onClick={modalRef.current?.close}
+                >
+                    <MaterialIconUI>cancel</MaterialIconUI>
+                </div>
+            </div>
+            <div className={styles.ModalContent}>
+                <img src={props.attachmentUrl} alt="Preview" />
```
Comment (naturetime10) [inline src/components/internal/PreviewAttachment.component.tsx]: Do we want make preview image only or any attachment?

File: src/libs/ui/AttachmentList.tsx [hunk 9]
```diff
@@ -30,13 +31,17 @@ export const AttachmentList: React.FC<Props> = (props) => {
...
+    };
+
     return (
         <>
             <div className={styles.Attachments}>
                 {props.attachments?.map((attachment) => {
                     return (
                         <div className={styles.Attachment} key={attachment.id}>
-                            <div className={styles.LeftSection}>
+                            <div className={styles.LeftSection} onClick={handlePreviewAttachment(attachment.url)}>
```
Comment (naturetime10) [inline src/libs/ui/AttachmentList.tsx]:
```suggestion
                            <div className={styles.LeftSection} 
                            onClick={handlePreviewAttachment(attachment.url, attachment.type)}>
```

File: src/libs/ui/MarkdownEditor.tsx [hunk 10]
```diff
@@ -47,6 +47,7 @@ interface Props {
         url: string;
     }>;
     onContentUpdated: () => void;
+    onPreviewAttachment: (url: string) => void;
```
Comment (naturetime10) [inline src/libs/ui/MarkdownEditor.tsx]:
```suggestion
    onPreviewAttachment: (attachmentType: AttachmentType, url: string) => void;
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Comment by naturetime10** [comment]
> will we create button to allow user to go next and prev previews?

https://fonts.google.com/icons?selected=Material+Symbols+Outlined:arrow_back_ios:FILL@1;wght@400;GRAD@0;opsz@24&icon.size=24&icon.color=%235f6368

https://fonts.google.com/icons?selected=Material+Symbols+Outlined:arrow_back_ios:FILL@1;wght@400;GRAD@0;opsz@24&icon.size=24&icon.color=%235f6368

File: src/components/internal/Modals/TaskDetailModal.component.tsx [hunk 11]
```diff
@@ -76,7 +79,6 @@ export class TaskDetailModalComponent extends Component<Props, State> {
             isControlKeyPressed: false,
         };
     }
```
Comment (naturetime10) [inline src/components/internal/Modals/TaskDetailModal.component.tsx]:
```suggestion
    }
    
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/libs/ui/AttachmentList.tsx [hunk 12]
```diff
@@ -10,6 +10,7 @@ interface Props {
     relativeLayout: RelativeLayout;
     onDeleteAttachment: (attachmentId: number) => void;
     onCopyAttachmentLink: (url: string) => void;
+    onPreviewAttachment: (url: string) => void;
```
Comment (naturetime10) [inline src/libs/ui/AttachmentList.tsx]:
```suggestion
    relativeLayout: RelativeLayout;
    onDeleteAttachment?: (attachmentId: number) => void;
    onCopyAttachmentLink?: (url: string) => void;
    onPreviewAttachment?: (url: string) => void;
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Comment by Travisliang001** [comment]
> will we create button to allow user to go next and prev previews?

will create separate PR for this

File: src/components/internal/Modals/PreviewAttachmentModal.component.tsx [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    previewAttachmentType?: string;
+}
+
+export class PreviewAttachemntModalComponent extends Component<Props, State> {
+    private readonly modalRef = createRef<ModalUI>();
+    constructor(props: Props) {
+        super(props);
+        this.state = {
+            previewAttachmentUrl: '',
+            previewAttachmentType: '',
```
Comment (naturetime10) [inline src/components/internal/Modals/PreviewAttachmentModal.component.tsx]:
```suggestion
```

File: src/components/internal/Modals/PreviewAttachmentModal.component.tsx [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+interface Props {}
+
+interface State {
+    previewAttachmentUrl?: string;
+    previewAttachmentType?: string;
+}
+
+export class PreviewAttachemntModalComponent extends Component<Props, State> {
+    private readonly modalRef = createRef<ModalUI>();
+    constructor(props: Props) {
```
Comment (naturetime10) [inline src/components/internal/Modals/PreviewAttachmentModal.component.tsx:16]:
```suggestion

    constructor(props: Props) {
```

File: src/components/internal/Modals/PreviewAttachmentModal.component.tsx [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+                    >
+                        <MaterialIconUI>cancel</MaterialIconUI>
+                    </div>
+                </div>
+                <div className={styles.PreviewAttachmentContainer}>
+                    <img
+                        src={this.state.previewAttachmentUrl}
+                        className={styles.PreviewImage}
+                        alt='Preview'
+                    />
```
Comment (naturetime10) [inline src/components/internal/Modals/PreviewAttachmentModal.component.tsx]: Can we extract this to a separate function and put switch case inside and switch on `previewAttachmentType`?

File: src/libs/ui/AttachmentList.module.scss [hunk 16]
```diff
@@ -22,7 +22,7 @@
                 height: 80px;
                 width: 80px;
                 border-radius: 4px;
-
+                cursor: pointer;
```
Comment (naturetime10) [inline src/libs/ui/AttachmentList.module.scss:25]:
```suggestion
                cursor: pointer;
                
```

File: src/components/internal/Modals/PreviewAttachmentModal.component.module.scss [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+        margin: 10px auto 0 auto;
+        width: 120px;
+    }
+}
+
+.PreviewAttachmentContainer {
+    width: 50vw;
+    height: 50vh;
+    min-width: 300px; /* Minimum width for the container */
+    min-height: 300px; /* Minimum height for the container */
```
Comment (naturetime10) [inline src/components/internal/Modals/PreviewAttachmentModal.component.module.scss]: Just curious, what's the reason for for min width and min hight? Is this for modal title and close button to be render without issue?

File: src/components/internal/Modals/PreviewAttachmentModal.component.module.scss [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+    overflow: auto; /* Enable scrolling */
```
Comment (naturetime10) [inline src/components/internal/Modals/PreviewAttachmentModal.component.module.scss]:
```suggestion
    overflow: auto;
```

File: src/libs/ui/AttachmentList.tsx [hunk 19]
```diff
@@ -23,11 +28,22 @@ export const AttachmentList: React.FC<Props> = (props) => {
     };
 
     const handleDeleteAttachment = (attachmentId: number) => () => {
-        props.onDeleteAttachment(attachmentId);
+        if (props.onDeleteAttachment) {
+            props.onDeleteAttachment(attachmentId);
+        }
```
Comment (naturetime10) [inline src/libs/ui/AttachmentList.tsx]:
```suggestion
        props.onDeleteAttachment?.call(null, attachmentId);
```

File: src/libs/ui/AttachmentList.tsx [hunk 20]
```diff
@@ -23,11 +28,22 @@ export const AttachmentList: React.FC<Props> = (props) => {
...
     };
 
     const handleCopyAttachmentLink = (url: string) => () => {
-        props.onCopyAttachmentLink(url);
+        if (props.onCopyAttachmentLink) {
+            props.onCopyAttachmentLink(url);
+        }
```
Comment (naturetime10) [inline src/libs/ui/AttachmentList.tsx]:
```suggestion
        props.onCopyAttachmentLink?.call(null, url);
```

**Review by naturetime10 — APPROVED** [review]

File: src/components/internal/Modals/PreviewAttachmentModal.component.module.scss (hunk 17 above)
Comment (Travisliang001) [reply src/components/internal/Modals/PreviewAttachmentModal.component.module.scss]: I was thinking if the image inside is smaller than the specified dimensions, the container itself will collapse. Therefore, I decided to use min-height and min-width.. but removing them and seems to work as well.

Comment (naturetime10) [reply src/components/internal/Modals/PreviewAttachmentModal.component.module.scss]: @Travisliang001 We can keep the `min-width` and `min-height`. We may need to center the image in the container if the image is smaller than `min-width` and `min-height`.

File: src/components/internal/Modals/PreviewAttachmentModal.component.tsx [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+    previewAttachmentType?: string;
+}
+
+export class PreviewAttachemntModalComponent extends Component<Props, State> {
+    private readonly modalRef = createRef<ModalUI>();
+
+    constructor(props: Props) {
+        super(props);
+        this.state = {
+        };
```
Comment (naturetime10) [inline src/components/internal/Modals/PreviewAttachmentModal.component.tsx]:
```suggestion
        this.state = {};
```

File: src/components/internal/Modals/PreviewAttachmentModal.component.tsx [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+        this.modalRef.current?.open();
+    };
+
+    private onCloseButtonClick = () => {
+        this.modalRef.current?.close();
+    };
+
+    private renderPreviewContent() {
+        const { previewAttachmentUrl, previewAttachmentType } = this.state;
+
```
Comment (naturetime10) [inline src/components/internal/Modals/PreviewAttachmentModal.component.tsx]:
```suggestion
```

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**